docs/ABI.rst

```diff
@@ -1005,181 +1060,187 @@ mangled in to disambiguate.
+  OPERATOR-CHAR ::= 'e'                      // = 'equals'
+  OPERATOR-CHAR ::= 'g'                      // > 'greater'
+  OPERATOR-CHAR ::= 'l'                      // < 'less'
+  OPERATOR-CHAR ::= 'm'                      // * 'multiply'
+  OPERATOR-CHAR ::= 'n'                      // ! 'not'
+  OPERATOR-CHAR ::= 'o'                      // | 'or'
+  OPERATOR-CHAR ::= 'p'                      // + 'plus'
+  OPERATOR-CHAR ::= 'q'                      // ? 'question'
+  OPERATOR-CHAR ::= 'r'                      // % 'remainder'
+  OPERATOR-CHAR ::= 's'                      // - 'subtract'
+  OPERATOR-CHAR ::= 't'                      // ~ 'tilde'
+  OPERATOR-CHAR ::= 'x'                      // ^ 'xor'
+  OPERATOR-CHAR ::= 'z'                      // . 'zperiod'
+
+  identifier ::= '0' ( [0-9]* 'a-z' )* [A-Z] [0-9]* IDENTIFIER-STRING    // identifier with word substitutions
+
+If the run-length start with a ``0`` it the identifier string contains
+word substitutions. A word is a sub-string of an identifier.
+A word starts either with an uppercase letter ``[A-Z]`` if the preceeding
+character is not an uppercase letter. Or it starts with a lowercase letter
+``[a-z]`` if the preceeding character is not a letter ``[a-zA-Z]``.
+A word ends either with a lowercase letter if the next character is not
+a lowercase letter. Or it ends with an uppercase letter if the next character
+is not a letter::
+
+  AbcDefGHI         // contains three words 'Abc', 'Def' and GHI
+  _abc_def_Ghi      // contains three words 'abc', 'def' and Ghi
+
+The words of all identifiers, which are encoded in the current mangling are
+enumerated and assigned to a letter: a = first word, b = second word, etc.
+
+The run-length of an identifier with word substitutions is a sequence of
+numbers and word-references, whereas all but the last word-reference are
+lowercase letters and the last one is an uppercase letter. Let's assume the
+current mangling already encoded the identifier ``AbcDefGHI``::
+
+  02ac1BMy_    // expands to: MyAbcGHI_Def
+
+A maximum of 26 words in a mangling can be used for substitutions.
+
+::
+
+  identifier ::= '00' natural identifier-start-char identifier-char*
+  identifier ::= '00' 'o' operator-fixity natural identifier-char*
 
 Identifiers that contain non-ASCII characters are encoded using the Punycode
 algorithm specified in RFC 3492, with the modifications that ``_`` is used
 as the encoding delimiter, and uppercase letters A through J are used in place
 of digits 0 through 9 in the encoding character set. The mangling then
-consists of an ``X`` followed by the run length of the encoded string and the
+consists of an ``00`` followed by the run length of the encoded string and the
 encoded string itself. For example, the identifier ``vergüenza`` is mangled
-to ``X12vergenza_JFa``. (The encoding in standard Punycode would be
+to ``0012vergenza_JFa``. (The encoding in standard Punycode would be
 ``vergenza-95a``)
```

Comment:
There's a bug with this mangling scheme for Unicode we need to fix: https://bugs.swift.org/browse/SR-660 If the first ASCII character in an identifier is a digit, as in `é22`, then we'll end up butting the `22` against the run length of the encoded string. We could fix that by just inserting a `_`, or encoding UTF-8 directly (though Punycode should in theory be more efficient for single-script identifiers).